packer/amazon/centos-k8s-worker.json

```diff
@@ -0,0 +1,87 @@
+{
+  "variables": {
+    "aws_access_key": "",
+    "aws_secret_key": "",
+    "region": "{{env `AWS_DEFAULT_REGION`}}",
+    "tarmak_environment": "{{env `TARMAK_ENVIRONMENT`}}",
+    "tarmak_base_image_name": "{{env `TARMAK_BASE_IMAGE_NAME`}}",
+    "puppet_path": "{{env `PUPPET_PATH`}}",
+    "ebs_volume_encrypted": "{{env `EBS_VOLUME_ENCRYPTED`}}"
+  },
+  "builders": [
+    {
+      "type": "amazon-ebs",
+      "access_key": "{{user `aws_access_key`}}",
+      "secret_key": "{{user `aws_secret_key`}}",
+      "region": "{{user `region`}}",
+      "source_ami_filter": {
+        "filters": {
+          "owner-alias": "aws-marketplace",
+          "product-code": "aw0evgkw8e5c1q413zgy5pjce",
+          "virtualization-type": "hvm"
+        },
+        "most_recent": true
+      },
+      "instance_type": "c4.xlarge",
+      "ssh_username": "centos",
+      "ssh_pty": "true",
+      "encrypt_boot": "{{user `ebs_volume_encrypted`}}",
+      "ami_name": "Tarmak CentOS 7 x86_64 with puppet-agent optimized for k8s workers {{isotime \"2006-01-02_030405\"}}",
+      "tags": {
+        "Name": "tarmak_{{user `tarmak_environment`}}_{{user `tarmak_base_image_name`}}",
+        "tarmak_environment": "{{user `tarmak_environment`}}",
+        "tarmak_base_image_name": "{{user `tarmak_base_image_name`}}"
+      },
+      "associate_public_ip_address": true
+    }
+  ],
+  "provisioners": [
+    {
+      "type": "shell",
+      "execute_command": "sudo -E -S sh '{{ .Path }}'",
+      "inline": [
+        "echo $PWD",
```

Comment:
Running into weird behaviour when I execute `tarmak cluster images build`. The packer build seems to delete the first line. Because of this it didn't add the Puppet repo and Puppet couldn't be installed. I now added an echo as the first line and it works. 

I tried replicating this with Packer (same version as Tarmak Packer) directly and I wasn't able to.

Going to investigate further now I know what the problem is.

Reply:
At a certain moment Packer creates this file locally and there it is already wrong. This means it goes wrong somewhere on the system where Packer is run and not on the instance itself.

```
INFO[0035] 1543309677,,ui,say,==> amazon-ebs: Provisioning with shell script: /var/folders/1h/11bqph8n0wl43my28ptqdp7m0000gn/T/packer-shell927284083  app=tarmak module=packer region=eu-west-1 tarmak_base_image_name=centos-k8s-worker tarmak_environment=singlecluster
2018/11/27 09:07:57 Opening /var/folders/1h/11bqph8n0wl43my28ptqdp7m0000gn/T/packer-shell927284083 for reading
2018/11/27 09:07:57 [DEBUG] Opening new ssh session
2018/11/27 09:07:57 [DEBUG] Starting remote scp process:  scp -vt /tmp
2018/11/27 09:07:57 [DEBUG] Started SCP session, beginning transfers...
2018/11/27 09:07:57 [DEBUG] Copying input data into temporary file so we can read the length
2018/11/27 09:07:57 [DEBUG] scp: Uploading script_1362.sh: perms=C0644 size=765
2018/11/27 09:07:57 [DEBUG] SCP session complete, closing stdin pipe.
2018/11/27 09:07:57 [DEBUG] Waiting for SSH session to complete.
2018/11/27 09:07:57 [DEBUG] scp stderr (length 31): Sink: C0644 765 script_1362.sh
2018/11/27 09:07:57 [DEBUG] Opening new ssh session
2018/11/27 09:07:57 [DEBUG] starting remote command: chmod 0755 /tmp/script_1362.sh
2018/11/27 09:07:57 [DEBUG] Opening new ssh session
2018/11/27 09:07:57 [DEBUG] starting remote command: sudo -E -S sh '/tmp/script_1362.sh'
```


output of that file is:
```
#!/bin/sh -e
rm -f /etc/vault/init-token
rpm -ivh https://yum.puppetlabs.com/puppetlabs-release-pc1-el-7.noarch.rpm
yum update -y
yum install -y epel-release
yum install -y git puppet-agent vim tmux socat python-pip at jq unzip awscli
sed -i '/GRUB_CMDLINE_LINUX=/c\GRUB_CMDLINE_LINUX="console=tty0 crashkernel=0 console=ttyS0,115200 biosdevname=0 net.ifnames=0"' /etc/sysconfig/grub
sed -i '/GRUB_CMDLINE_LINUX=/c\GRUB_CMDLINE_LINUX="console=tty0 crashkernel=0 console=ttyS0,115200 biosdevname=0 net.ifnames=0"' /etc/default/grub
grub2-mkconfig -o /boot/grub2/grub.cfg
rm -f /etc/sysconfig/network-scripts/ifcfg-ens*
systemctl disable kdump.service
rm -f /etc/ssh/ssh_host_*
mkdir -p /tmp/packer-puppet-masterless
chown centos:centos /tmp/packer-puppet-masterless
```

Reply:
It goes wrong in here https://github.com/jetstack/tarmak/blob/012e66c8c0e911c494b5478b6812f5f77cdf58f4/vendor/github.com/hashicorp/packer/provisioner/shell/provisioner.go#L90